### ISSUE TITLE
Kotlin Multiplaform Mobile SDK

### ISSUE THREAD

**Issue by IanArb** [thenewboston-developers/Project-Proposals#169]
#### How would you describe your project?
We've already started implementing all the core API endpoints for the Kotlin SDK repository here - https://github.com/thenewboston-developers/Kotlin-SDK

Kotlin Multiplatform Mobile is a framework introduced by Jetbrains to share core business logic across platforms such as Android and iOS. Instead of using cross platform technologies such as React Native or Flutter we can share the business logic and leverage the native capabilities of Android and iOS for better user experience. 

This means we won't have to have 2 projects for each platform, Android and iOS. The core business logic would be written in Kotlin and compiled to the specific platform’s language, Swift/Objective C for iOS and Java/Kotlin for Android. I encourage you to read the official statement and the use cases from Jetbrains. https://kotlinlang.org/lp/mobile/

#### What problem does it solve?
The goal of the Kotlin Multiplatform SDK is to provide a wrapper library for the REST APIs. This will target Android and iOS platforms. It will release an aar library for Android and a cocoapods framework for iOS. 

#### Who is the target market?
Android and iOS apps that wish to integrate the TNB crypto currency into their products.

#### What do you think is the opportunity of this project?
The opportunity to provide SDKs to mobile platforms and unlock our crypto currency in the mobile space. For instance, exchanges such as Coinbase already have mobile SDKs and an app.

#### How does it integrate with thenewboston digital currency work?
It provides the network calls to the API endpoints in the currency network and exposes them as a wrapper library. This reduces the workload for an app to integrate the API endpoints where it would just need to call the repository for the specific API. e.g. ConfirmationRepository.kt in the SDK implementation to date. 

Example of repository - https://github.com/thenewboston-developers/Kotlin-SDK/blob/master/lib/src/main/java/com/thenewboston/api/confirmationvalidatorapi/repository/ConfirmationRepository.kt

#### Are there any competitors?
Yes. Any crypto network that have mobile SDKs for Android and iOS would be considered a competitor. 

#### Will you need specific information or advice?
Yes, we would need a product manager to define a roadmap for the SDKs. The reason being is that we are providing a wrapper library for any potential mobile app integration. The SDKs will need to be closely aligned with the backend API roadmap.

#### What are your goals and objectives?
To provide SDKs for both Android and iOS that want to integrate the TNB network currency into their own mobile app products. Instead of creating an SDK for each platform, we can create it once by using Kotlin Multiplatform technology that can target both iOS and Android ecosystems. This is possible by publishing .aar artifacts for Android and framework for iOS using cocoapods. 

#### How will you evaluate project success?
The big way to measure success would be how many apps have successfully integrated the SDK into their products.

#### What risks should you plan to manage?
The current risk is that the concurrency model of Kotlin Multiplatform is not stable. There is ongoing work to improve interoperability on the concurrency model in Kotlin Native compiler which is what is used to compile to iOS.

More information can be found here --> https://blog.jetbrains.com/kotlin/2020/07/kotlin-native-memory-management-roadmap/

#### What are your milestones?

- Milestone 1 : API endpoints implemented in Kotlin - Already completed. See [https://github.com/thenewboston-developers/Kotlin-SDK](https://github.com/thenewboston-developers/Kotlin-SDK)
- Milestone 2 : Test the SDK with an Android sample app integration - May 1st
- Milestone 3: Port the Kotlin SDK to multiplatform - June 1st
- Milestone 4: Integrate into iOS sample app - June 25th
- Milestone 5: Release beta SDKs for Android and iOS - August 1st
- Milestone 6: Test the SDKs in any mobile app - August 15th

#### Tell us about your team members:
Full Name, Slack Username, GitHub Username, LinkedIn, Role, Team Joining Date, Team Leaving Date, Account number

Current team members can be found here --> https://thenewboston.com/teams/Kotlin-SDK/Members

![image](https://user-images.githubusercontent.com/6011914/111692913-e6126780-8827-11eb-9a71-f8730a02a6cd.png)

#### List the roles of any additional team members you would be looking for?
x1 Product Manager
x1 iOS Developer (Must have experience with iOS frameworks and cocoapods) 
x1 Android Developer (Must have experience in SDKs and KMM)

**Comment by manishram** [comment]
Hey, @IanArb please update your proposal with a new proposal format and details.
Check the new project proposal format [HERE](https://github.com/thenewboston-developers/Projects/issues/new?assignees=&labels=Project&template=project-proposal.md&title=NAME_OF_YOUR_PROJECT)

**Comment by IanArb** [comment]
Hey @manishram updated with the new format.

**Comment by manishram** [comment]
Govs. are going to decide if this proposal should be treated as a project or merge to the core project.

**Comment by angle943** [comment]
@IanArb hey Ian, a couple things are unclear to me.

For your milestone 1, it says that you will integrate the SDK into a sample android app. A couple questions:

1) is the SDK done for you to start integrating it?
2) what is the sample android app?
3) Has the sample android app been completed (albeit minus the sdk integration), or are you planning on building it from scratch?

**Comment by IanArb** [comment]
Hey @angle943 Justin,

1. Correct, the SDK development is complete. It needs to be tested for integration into the sample app. Then if there are no bugs or issues we can create the library package. 
2. The sample Android app is just a simple UI like the account manager for testing the API calls from the SDK
3. Yes the sample app is ready and just needs the SDK to be integrated into it.

**Comment by angle943** [comment]
FYI these questions refer strictly to milestone 1:

1. Will you be publishing the sample android app anywhere?
2. what will be the proof of work for the sample app sdk integration? (ie how would the auditors be able to check that you have completed milestone 1)

**Comment by IanArb** [comment]
> 1. Will you be publishing the sample android app anywhere?
The sample app lives the repository and will be only used for internal usage for the SDK integration. For example, we develop a new endpoint and we can use the sample app to validate these changes.

> 2. what will be the proof of work for the sample app sdk integration? (ie how would the auditors be able to check that you have completed milestone 1)
You can validate the proof of work via the Pull Requests and the app module in the repository. 

> 3. what functionalities exactly does the sample app have? It is hard for us to guage how long the integration will take & the value in coins it will be worth without knowing exactly what functionalities the app has, and what functions you will be porting over to use the Kotlin sdk
The functionality of the sample app consists of touchpoints as follows:

Account overview for transactions and balance
Banks and Validators overview for transactions, accounts, blocks, confirmation blocks etc..

Basically any endpoint touchpoint from the API is covered in the sample app.

**Comment by angle943** [comment]
Milestone 1 has been approved by the governors. Specifically, the amount of 12,500. Once that is completed, please write up a `milestone payout request` for the work done, and then we can discuss the next milestone (the one for 27,500)

**Comment by IanArb** [comment]
Hey @angle943 thanks for that. Just saw this now.

**Comment by manishram** [comment]
Hey @IanArb please comment here your project repo url.

**Comment by IanArb** [comment]
Hey @manishram https://github.com/thenewboston-developers/Kotlin-SDK
